### PR TITLE
Login Routing & Access

### PR DESCRIPTION
**Pull request by monocliffe**
[TPOZ-1517](https://wpengine.atlassian.net/browse/TPOZ-1517)
[TPOZ-1521](https://wpengine.atlassian.net/browse/TPOZ-1521)
[TPOZ-1522](https://wpengine.atlassian.net/browse/TPOZ-1522)

When requesting a restricted page, specified in the controller, the user is redirected to the login page. After logging in the user is redirected back to the originally requested page.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,11 +1,23 @@
 class ApplicationController < ActionController::Base
-
+  protect_from_forgery with: :exception
   before_action :current_user
+  #before_action :ensure_login, except: [:index, :new]
```
Comment (JamesDominy) [inline app/controllers/application_controller.rb]: Try not to commit commented code ... `git-cola` and `git add -p` are your friends

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 
   protected
 
   def current_user
     return unless session[:user_id]
     @current_user = User.where(id: session[:user_id]).first
   end
+
+  def logged_in?
+    !@current_user.nil?
```
Comment (JamesDominy) [inline app/controllers/application_controller.rb]: In general, positive logic assertions are easier to read, e.g. `@current_user.present?`. This is especially true with 'double negatives' arising from assertions about `nil`

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,11 +1,23 @@
...
+  end
+  helper_method :logged_in?
+
+  def ensure_login
+    return true if logged_in?
+    session[:return_to] = request.fullpath
+    redirect_to login_path and return false
```
Comment (JamesDominy) [inline app/controllers/application_controller.rb]: I don't think you need the `and return` in this case, see https://stackoverflow.com/questions/34842427/redirect-to-vs-redirect-to-and-return

**Review by JamesDominy — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (monocliffe) [reply app/controllers/application_controller.rb]: Reading back over it, I agree. Changed.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (monocliffe) [reply app/controllers/application_controller.rb]: Also true. Useful to know.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (monocliffe) [reply app/controllers/application_controller.rb]: I'll look into them both. Resources about `git add -p` are kinda sparse, do you have any links?

Comment (monocliffe) [reply app/controllers/application_controller.rb]: I removed the line in this case as it was no longer useful.

**Review by JamesDominy — APPROVED** [review]
👍

**Pull request merged**